ItemChanger/Modules/ProgressiveItemGroupModule.cs

```diff
@@ -0,0 +1,181 @@
+            OrderedTransitivePredecessorsLookup[item.name] = tag.OrderedTransitivePredecessors;
+            item.ModifyItem += ModifyItem;
+        }
+
+        /// <summary>
+        /// Modifies the given item according to the output of <see cref="GetActualItems(IEnumerable{string}, List{string}, Dictionary{string, List{string}})"/>.
+        /// </summary>
+        protected void ModifyItem(GiveEventArgs args)
+        {
+            Dictionary<string, int> prev = GetActualItems(CollectedItemList, OrderedMemberList, OrderedTransitivePredecessorsLookup);
+            CollectedItemList.Add(args.Orig.name);
+            Dictionary<string, int> next = GetActualItems(CollectedItemList, OrderedMemberList, OrderedTransitivePredecessorsLookup);
+
+            string nextItem = next.Single(kvp => kvp.Value > (prev.TryGetValue(kvp.Key, out int value) ? value : 0)).Key;
+            args.Item = GroupItems[nextItem];
+        }
+
+        /// <summary>
+        /// Gets the multiset of items which replaces the sequence of collected items.
+        /// The sequence is first resorted according to the order of the main item list. Then each item is replaced by its first predecessor not yet in the result, or else unmodified.
+        /// If an item is added when it and all of its predecessors are given, its multiplicity in the dictionary may be incremented beyond 1.
+        /// </summary>
+        public static Dictionary<string, int> GetActualItems(IEnumerable<string> collectedItems, List<string> itemList, Dictionary<string, List<string>> predecessors)
+        {
+            Dictionary<string, int> result = [];
+
+            foreach (string item in collectedItems.OrderBy(itemList.IndexOf))
+            {
+                bool replaced = false;
+                foreach (string p in predecessors[item])
+                {
+                    if (!result.ContainsKey(p))
+                    {
+                        result.Add(p, 1);
+                        replaced = true;
+                        break;
+                    }
+                }
+                if (!replaced)
+                {
+                    if (!result.ContainsKey(item))
+                    {
+                        result.Add(item, 1);
+                    }
+                    else
+                    {
+                        result[item]++;
+                    }
+                }
+            }
+            return result;
+        }
+
+        /// <summary>
+        /// Checks that the tag-provided data satisfies all compatibility requirements with each other and with the module.
+        /// In particular, all expected items must be registered with the module, 
+        /// the "predecessor" relation must be a strict partial order (transitive and irreflexive),
+        /// and this partial order must be consistent with the order on the module's <see cref="OrderedMemberList"/>.
+        /// </summary>
+        protected void LateInitialize()
```

Comment:
This approach seems a bit problematic for cases like Archipelago in which not every tagged item will necessarily be in the file before it is received, is there a way we can handle such scenarios?